torch/csrc/utils/tensor_new.cpp

```diff
@@ -488,6 +488,17 @@ Tensor legacy_tensor_ctor(c10::DispatchKey dispatch_key, at::ScalarType scalar_t
     at::OptionalDeviceGuard device_guard(deviceOptional);
     return at::empty({0}, options(dispatch_key, scalar_type));
   } else if (r.idx == 1) {
+    at::ScalarType storage_scalar_type =
+        reinterpret_cast<THPDtype*>(
+            PyObject_GetAttrString(r.pyobject(0), "dtype"))
```

Comment:
This leaks a python reference.

Reply:
I changed all of these leaks to this:
```
    PyObject* storage_pyobject = r.pyobject(0);
    at::ScalarType storage_scalar_type =
        reinterpret_cast<THPDtype*>(
            PyObject_GetAttrString(storage_pyobject, "dtype"))
            ->scalar_type;
    Py_DECREF(storage_pyobject);
```
That should be sufficient, right?

Reply:
It's `PyObject_GetAttrString` that causes the leak. It returns a new reference. I think `THPObjectPtr` can be used as an RAII wrapper:
```cpp
THPObjectPtr dtype_attr(PyObject_GetAttrString(r.pyobject(0), "dtype"));
if (!dtype_attr) throw python_error();
auto storage_scalar_type = static_cast<THPDtype*>(
    dtype_attr.get())->scalar_type;
```

Reply:
Cool thanks!